# Foundations of R Software — Week 9
## Strings: Display, Formatting & Manipulation

**Source:** NPTEL, Prof. Shalabh (IIT Kanpur), Lectures 36–39

| Lecture | Topic | Functions |
|---|---|---|
| 36 | Print and format with concatenation | `print`, `cat`, `format` |
| 37 | Paste function | `paste`, `paste0`, `sep`, `collapse` |
| 38 | String splitting | `strsplit` |
| 39 | Manipulating strings | `nchar`, `nzchar`, `tolower`, `toupper` |

> **Colab tip:** this notebook needs the **R** runtime. If your kernel says Python, open  
> `https://colab.research.google.com/notebook#create=true&language=r` and paste the cells in,  
> or go to *Runtime → Change runtime type* and choose **R**. No packages need installing — everything uses base R.

In [ ]:
# Sanity check: confirm we are running R
R.version.string

[1] "R version 4.6.1 (2026-06-24)"

---
# Lecture 36 — Print and Format with Concatenation

## 1. The limitation of `print()`
`print()` can show **only one object at a time**. Passing several items is an error.

Goal: display  **The zero occurs at 2*pi radians.**

In [ ]:
# This FAILS: print accepts one object. The 2nd argument is treated as `digits`, etc.
# tryCatch lets us show the error message without stopping the notebook.
tryCatch(
  print("The zero occurs at", 2*pi, "radians."),
  error = function(e) cat("Error:", conditionMessage(e), "\n")
)

Error: invalid 'quote' argument 


**Workaround:** print each piece separately — works, but the output is split over several lines and shows `[1]` prefixes and quotes.

In [ ]:
print("The zero occurs at"); print(2*pi); print("radians")

[1] "The zero occurs at"
[1] 6.283185
[1] "radians"


## 2. `cat()` — combine multiple items into continuous output

`cat()` converts its arguments to character vectors, links them in sequence, and puts a separator (`sep`, default a **space**) between each element.

**Usage**
```r
cat(..., file = "", sep = " ", fill = FALSE, labels = NULL, append = FALSE)
```

| Argument | Meaning |
|---|---|
| `sep` | string placed between items (default `" "`) |
| `fill` | `FALSE` (default): only explicit `"\n"` breaks lines. `TRUE`/number: auto-wrap at console width / that width |
| `labels` | character vector of labels for each printed line (used when `fill` is on) |
| `file`, `append` | write to a file; `append = TRUE` adds instead of overwriting |

⚠️ `cat()` **does not add a newline** — you must add `"\n"` yourself.

In [ ]:
cat("The zero occurs at", 2*pi, "radians.", "\n")

The zero occurs at 6.283185 radians. 


`cat()` is for *displaying* only — unlike `paste()` (Lecture 37), its result cannot be assigned to a variable.

### Example: current date

In [ ]:
d = date()
cat("Today's date is:", d, "\n")   # output depends on when you run it

Today's date is: Mon Sep 21 08:52:02 2026 


### Example: the `sep` argument

In [ ]:
x = 1:10
x

[1]  1  2  3  4  5  6  7  8  9 10

In [ ]:
cat(x, sep = " ++ ")
cat("\n")            # move to a new line
cat(x, sep = " / ")
cat("\n")

1 ++ 2 ++ 3 ++ 4 ++ 5 ++ 6 ++ 7 ++ 8 ++ 9 ++ 10
1 / 2 / 3 / 4 / 5 / 6 / 7 / 8 / 9 / 10


### Example: mixing text and computed values

In [ ]:
x = 7
cat("The square of", x, "is", x^2, "!\n")

The square of 7 is 49 !


### Example: `format()` to control digits

In [ ]:
cat("The square root of", x, "is approximately", format(sqrt(x), digits = 3), "\n")

The square root of 7 is approximately 2.65 


### Example: `cat()` with vectors

In [ ]:
evenno = c(2, 4, 6, 8, 10)
evenno
cat("The first few even numbers are:", evenno, "...\n")

[1]  2  4  6  8 10

The first few even numbers are: 2 4 6 8 10 ...


### Example: `fill` and `labels`
With `fill = 2` (a very narrow width), every item goes on its own line and gets a label.

In [ ]:
x = 1:10
cat(x, fill = 2, labels = paste("(", letters[1:10], "):"))

( a ): 1 
( b ): 2 
( c ): 3 
( d ): 4 
( e ): 5 
( f ): 6 
( g ): 7 
( h ): 8 
( i ): 9 
( j ): 10


---
# Lecture 37 — The `paste()` Function

`paste()` concatenates strings by joining them end-to-end.

* Converts arguments to character (via `as.character`) and joins them.
* **Returns a value** → can be assigned to a variable (contrast with `cat()`).
* Vector arguments are concatenated **element-wise**, giving a character vector.

**Usage**
```r
paste(..., sep = " ", collapse = NULL)
paste0(..., collapse = NULL)      # same as paste(..., sep = "")
```

* `sep` — separator placed between the strings being joined (default is a single space).
* `collapse` — if given, the resulting vector is joined into **one single string**, using this separator.
* `paste0()` — shortcut for `sep = ""`, slightly more efficient.

## 1. Conversion to character

In [ ]:
paste(1:12)
as.character(1:12)   # alternative to paste for a single vector

[1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12"

[1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12"

## 2. The `sep` argument

In [ ]:
paste("Everybody", "loves", "R Programming.")
paste("Everybody", "loves", "R Programming.", sep = "*")
paste("Everybody", "loves", "R Programming.", sep = "===")

[1] "Everybody loves R Programming."

[1] "Everybody*loves*R Programming."

[1] "Everybody===loves===R Programming."

## 3. Vector arguments (recycled element-wise)
Each element of `names` is combined with the other strings.

In [ ]:
names = c("Prof. Singh", "Mr.  Venkat", "Dr. Jha")
names
paste(names, "is", "a good", "person.")

[1] "Prof. Singh" "Mr.  Venkat" "Dr. Jha"

[1] "Prof. Singh is a good person." "Mr.  Venkat is a good person."
[3] "Dr. Jha is a good person."

## 4. The `collapse` argument
`collapse` is a *top-level* separator: it glues the generated strings into a **single** string.

In [ ]:
paste(names, "is", "a good", "person.", collapse = ", and ")

[1] "Prof. Singh is a good person., and Mr.  Venkat is a good person., and Dr. Jha is a good person."

## 5. Role of blank spaces inside quotes
Spaces inside the quotes are part of the string, so they are kept exactly as typed.

In [ ]:
paste(1, " is first", 2, " is second", 3, " is third", sep = "#")

[1] "1# is first#2# is second#3# is third"

In [ ]:
paste(1, " is first    ", 2, " is second    ", 3, " is third    ", sep = "#")

[1] "1# is first    #2# is second    #3# is third    "

## 6. Building labels: `sep` vs `collapse`

In [ ]:
x = paste("Ex", 1:5, sep = "_")
x
x[1]; x[2]; x[3]; x[4]; x[5]

[1] "Ex_1" "Ex_2" "Ex_3" "Ex_4" "Ex_5"

[1] "Ex_1"

[1] "Ex_2"

[1] "Ex_3"

[1] "Ex_4"

[1] "Ex_5"

In [ ]:
# With collapse, a single string is returned instead of a vector of 5
x = paste("Ex", 1:5, sep = "_", collapse = "")
x

[1] "Ex_1Ex_2Ex_3Ex_4Ex_5"

**Side by side — note the difference:**

In [ ]:
paste("Ex", 1:5, sep = "_")                    # vector of 5 strings
paste("Ex", 1:5, sep = "_", collapse = "")     # ONE string
length(paste("Ex", 1:5, sep = "_"))
length(paste("Ex", 1:5, sep = "_", collapse = ""))

[1] "Ex_1" "Ex_2" "Ex_3" "Ex_4" "Ex_5"

[1] "Ex_1Ex_2Ex_3Ex_4Ex_5"

[1] 5

[1] 1

## 7. `paste0()` vs `paste()`
With a **single vector**, both give the same result:

In [ ]:
paste0(1:10)
paste(1:10)

[1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10"

[1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10"

With **more than one vector**, `paste0` joins with *no* space, in a vectorised way (useful for ordinal suffixes).

In [ ]:
paste0(1:10, c("st", "nd", "rd", rep("th", 7)))

[1] "1st"  "2nd"  "3rd"  "4th"  "5th"  "6th"  "7th"  "8th"  "9th"  "10th"

In [ ]:
# Note the extra blank space that paste() inserts
paste(1:10, c("st", "nd", "rd", rep("th", 7)))

[1] "1 st"  "2 nd"  "3 rd"  "4 th"  "5 th"  "6 th"  "7 th"  "8 th"  "9 th" 
[10] "10 th"

---
# Lecture 38 — String Splitting with `strsplit()`

`strsplit()` splits the elements of a character vector into **substrings**, according to matches of `split`.  
`split` can be a single character or a character string.

**Usage**
```r
strsplit(x, split, fixed = FALSE, ...)
```

| Argument | Meaning |
|---|---|
| `x` | character vector, each element of which is split |
| `split` | pattern to split on — a **regular expression** unless `fixed = TRUE` |
| `fixed` | `TRUE`: match `split` exactly; `FALSE` (default): treat as regex |

The result is always a **list** — one element per element of `x`.

In [ ]:
x = "The&!syntax&!of&!paste&!is!&available! &inthe online-help"
x

[1] "The&!syntax&!of&!paste&!is!&available! &inthe online-help"

### Split on a single character `"!"`

In [ ]:
strsplit(x, split = "!")

[[1]]
[1] "The&"                "syntax&"             "of&"                
[4] "paste&"              "is"                  "&available"         
[7] " &inthe online-help"

### Split on the two-character string `"&!"`

In [ ]:
strsplit(x, split = "&!")

[[1]]
[1] "The"                               "syntax"                           
[3] "of"                                "paste"                            
[5] "is!&available! &inthe online-help"

### Split on `"!&"` and access individual components
Because the result is a list, use `[[1]]` first, then `[i]`.

In [ ]:
y = strsplit(x, split = "!&")
y

[[1]]
[1] "The&!syntax&!of&!paste&!is"    "available! &inthe online-help"

In [ ]:
y[[1]][1]
y[[1]][2]
y[[1]][3]     # only 2 pieces exist, so this gives NA

[1] "The&!syntax&!of&!paste&!is"

[1] "available! &inthe online-help"

[1] NA

### Example: dates → matrix
Split ISO dates into year / month / day, then arrange them in a matrix.

In [ ]:
dates = c("2020-07-24", "2021-08-25", "2022-09-26", "2023-10-27")
datesplt = strsplit(dates, "-")
datesplt

[[1]]
[1] "2020" "07"   "24"  

[[2]]
[1] "2021" "08"   "25"  

[[3]]
[1] "2022" "09"   "26"  

[[4]]
[1] "2023" "10"   "27"

`unlist()` flattens the list into one character vector; `matrix(..., byrow = TRUE)` then fills it row by row → a **character** matrix.

In [ ]:
datemat = matrix(unlist(datesplt), nrow = 4, ncol = 3, byrow = TRUE)
datemat

2020,07,24
2021,08,25
2022,09,26
2023,10,27


Wrap with `as.numeric()` to get a **numeric** matrix (note `"07"` becomes `7`).

In [ ]:
datematrix = matrix(as.numeric(unlist(datesplt)), nrow = 4, ncol = 3, byrow = TRUE)
datematrix

2020,7,24
2021,8,25
2022,9,26
2023,10,27


### Example: split a word into characters
Using `split = ""` breaks a string into single characters.

In [ ]:
strsplit("Shalabh", split = "")

[[1]]
[1] "S" "h" "a" "l" "a" "b" "h"

---
# Lecture 39 — Manipulating Strings & Alphabets

## 1. Counting characters: `nchar()` and `nzchar()`

* `nchar(x)` — returns the **number of characters** in each element of `x`.
* `nzchar(x)` — quickly tests whether each element is a **non-empty** string (`TRUE`/`FALSE`).

**Usage**
```r
nchar(x, type = "chars", allowNA = FALSE, keepNA = NA)
nzchar(x, keepNA = FALSE)
```
`type` is one of `"bytes"`, `"chars"`, `"width"`.

### `nchar()` on strings

In [ ]:
x = "R course 24.07.2022"
y = "Number of participants: 25"
nchar(x)   # count the number of characters
nchar(y)

[1] 19

[1] 26

### `nzchar()`

In [ ]:
nzchar(x)
nzchar(y)

[1] TRUE

[1] TRUE

In [ ]:
x = c("Apple", "Banana", "Cake")
x
nzchar(x)

[1] "Apple"  "Banana" "Cake"

[1] TRUE TRUE TRUE

In [ ]:
y = c("Apple", "", "Cake")     # middle element is an empty string
y
nzchar(y)

[1] "Apple" ""      "Cake"

[1]  TRUE FALSE  TRUE

### `nchar()` on a vector of strings and on numbers
Numbers are first coerced to character, so the decimal point counts as a character.

In [ ]:
x = c("Apple", "Banana", "Cake")
nchar(x)

y = c(2, 4, 6)
nchar(y)

z = c(11, 222, 3333)
nchar(z)

z1 = c(1.1, 2.22, 3.333)
nchar(z1)

[1] 5 6 4

[1] 1 1 1

[1] 2 3 4

[1] 3 4 5

## 2. Changing case: `tolower()` and `toupper()`

Convert upper-case characters to lower-case (or vice versa). **Non-alphabetic characters are left unchanged.**

In [ ]:
x = "R course will start from 24.07.2022"
toupper(x)

[1] "R COURSE WILL START FROM 24.07.2022"

In [ ]:
z = "INDIAN INSTITUTE OF TECHNOLOGY"
tolower(z)

[1] "indian institute of technology"

---
# Quick Reference

| Task | Function | Example | Result |
|---|---|---|---|
| Print many items on one line | `cat()` | `cat("a", 1, "\n")` | `a 1` |
| Custom separator in `cat` | `cat(sep=)` | `cat(1:3, sep=" / ")` | `1 / 2 / 3` |
| Limit digits | `format(digits=)` | `format(sqrt(7), digits=3)` | `"2.65"` |
| Join strings (returns value) | `paste()` | `paste("a","b")` | `"a b"` |
| Join with no separator | `paste0()` | `paste0("a","b")` | `"ab"` |
| Join into a single string | `collapse=` | `paste("Ex",1:3,collapse=", ")` | `"Ex 1, Ex 2, Ex 3"` |
| Split a string | `strsplit()` | `strsplit("a-b","-")` | list: `"a"` `"b"` |
| Number of characters | `nchar()` | `nchar("Cake")` | `4` |
| Non-empty test | `nzchar()` | `nzchar("")` | `FALSE` |
| To lower / upper case | `tolower()` / `toupper()` | `toupper("r")` | `"R"` |

### Key takeaways
1. `print()` takes **one** object; `cat()` takes **many** and doesn't add a newline.
2. `cat()` only displays; `paste()` **returns** a string you can store.
3. `sep` controls the separator *between arguments*; `collapse` joins the *whole result* into one string.
4. `strsplit()` returns a **list** — index with `[[i]][j]`, and use `unlist()` to flatten.

In [ ]:
# Bonus check of the quick-reference examples
cat("a", 1, "\n")
cat(1:3, sep = " / "); cat("\n")
format(sqrt(7), digits = 3)
paste("Ex", 1:3, collapse = ", ")
strsplit("a-b", "-")
nzchar("")

a 1 
1 / 2 / 3


[1] "2.65"

[1] "Ex 1, Ex 2, Ex 3"

[[1]]
[1] "a" "b"

[1] FALSE